Sources_v3/Workers/Background/MessageEditor.swift

```diff
@@ -57,58 +54,43 @@ class MessageEditor<ExtraData: ExtraDataTypes>: Worker {
         if wasEmpty {
-            database.backgroundReadOnlyContext.perform { [weak self] in
-                self?.processNextMessage()
-            }
+            processNextMessage()
         }
     }
 
     private func processNextMessage() {
         guard let messageId = pendingMessageIDs.first else { return }
         
-        guard let dto = database.backgroundReadOnlyContext.message(id: messageId) else {
-            removeMessageIDAndContinue(messageId)
-            return
-        }
-        
-        guard dto.localMessageState == .pendingSync else {
-            removeMessageIDAndContinue(messageId)
-            return
-        }
-        
-        markMessage(withID: messageId, as: .syncing) {
+        backgroundContext.perform { [weak self] in
+            guard let dto = self?.backgroundContext.message(id: messageId), dto.localMessageState == .pendingSync else {
+                self?.removeMessageIDAndContinue(messageId)
+                return
+            }
+            
             let requestBody = dto.asRequestBody() as MessageRequestBody<ExtraData>
-            self.apiClient.request(endpoint: .editMessage(payload: requestBody)) {
-                switch $0 {
-                case .success:
-                    self.markMessage(withID: messageId, as: nil) {
-                        self.removeMessageIDAndContinue(messageId)
-                    }
-                case .failure:
-                    self.markMessage(withID: messageId, as: .syncingFailed) {
-                        self.removeMessageIDAndContinue(messageId)
-                    }
+            
+            self?.markMessage(withID: messageId, as: .syncing)
+            self?.apiClient.request(endpoint: .editMessage(payload: requestBody)) { result in
+                self?.backgroundContext.perform {
+                    self?.markMessage(withID: messageId, as: result.error == nil ? nil : .syncingFailed)
+                    self?.removeMessageIDAndContinue(messageId)
                 }
             }
         }
     }
     
     private func removeMessageIDAndContinue(_ messageId: MessageId) {
         pendingMessageIDs.remove(messageId)
-        sendingQueue.async { self.processNextMessage() }
+        processNextMessage()
     }
     
-    private func markMessage(withID id: MessageId, as state: LocalMessageState?, completion: @escaping () -> Void) {
-        database.write({
-            $0.message(id: id)?.localMessageState = state
-        }, completion: {
-            if let error = $0 {
-                log.error("Error changing localMessageState for message with id \(id) to `\(String(describing: state))`: \(error)")
-            }
-            self.database.backgroundReadOnlyContext.perform {
-                completion()
-            }
-        })
+    private func markMessage(withID id: MessageId, as state: LocalMessageState?) {
+        backgroundContext.message(id: id)?.localMessageState = state
```

Comment:
We're using `backgroundReadOnlyContext` to mutate an object.. shouldn't we use `writableContext`?

Reply:
that's a very good point 🤔 But this breaks the linear flow 😄 will  update the implementation to use `writableContext`

Reply:
fixed in ded903840ae45a33977d9b77006517eb03188a57